src-input/duk_bi_encoding.c

```diff
@@ -197,15 +235,17 @@ DUK_INTERNAL duk_ret_t duk_bi_textencoder_prototype_encode(duk_context *ctx) {
 	duk_decode_string(ctx, 0, duk__utf8_encode_char, (void *) &enc_ctx);
 	if (enc_ctx.lead != 0x0000L) {
 		/* unpaired high surrogate at end of string */
-		enc_ctx.out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx.out);
+		DUK__EMIT_UTF8_REPL_CHAR(&enc_ctx);
 		DUK_ASSERT(enc_ctx.out <= output + (3 * len));
 	}
 
+	/* The output buffer is usually very much oversized, so shrink it to
+	 * actually needed size.
+	 */
+	DUK_ASSERT_TOP(ctx, 2);
+	duk_resize_buffer(ctx, -1, (duk_size_t) (enc_ctx.out - output));
```

Comment:
At first I thought this was an unnecessary step if bufferobjects are enabled.  But it's actually a memory hog if you don't since the plain buffer will be the concrete backing for the Uint8Array.